kotsadm/pkg/kurl/join_cert.go

```diff
@@ -0,0 +1,199 @@
+package kurl
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/replicatedhq/kots/kotsadm/pkg/logger"
+	"github.com/replicatedhq/kots/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+func createCertAndKey(ctx context.Context, client kubernetes.Interface, namespace string) (string, error) {
+	pod, err := getPodSpec(client, namespace)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create pod spec")
+	}
+
+	_, err = client.CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create pod")
+	}
+
+	defer func() {
+		go func() {
+			if err := client.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{}); err != nil {
+				logger.Errorf("Failed to delete pod %s: %v\n", pod.Name, err)
+			}
+		}()
+	}()
+
+	for {
+		status, err := client.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get pod")
+		}
+		if status.Status.Phase == corev1.PodRunning ||
+			status.Status.Phase == corev1.PodFailed ||
+			status.Status.Phase == corev1.PodSucceeded {
+			break
+		}
+		time.Sleep(time.Second * 1)
+	}
+
+	podLogs, err := getCertGenLogs(ctx, client, pod)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get pod logs")
+	}
+
+	key, err := parseCertGenOutput(podLogs)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to parse pod logs")
+	}
+
+	return key, nil
+}
+
+func getCertGenLogs(ctx context.Context, client kubernetes.Interface, pod *corev1.Pod) ([]byte, error) {
+	// maxLines := int64(10000)
+	podLogOpts := corev1.PodLogOptions{
+		Follow:    false,
+		Container: pod.Spec.Containers[0].Name,
+		// TailLines:
+		// SinceTime:
+	}
+
+	req := client.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &podLogOpts)
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get log stream")
+	}
+	defer podLogs.Close()
+
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, podLogs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to copy log")
+	}
+
+	return buf.Bytes(), nil
+}
+
+func parseCertGenOutput(logs []byte) (string, error) {
+	// Output looks like this:
```

Comment:
yikes, there's no JSON output or less brittle?

Reply:
it seems like the other option would be to write this in go and import kotsadm lib?

Reply:
@marccampbell @emosbaugh I kept this in its own commit intentionally.  We should figure out how to make this work, maybe in 1.19: https://github.com/replicatedhq/kots/commit/6ba2ca6a#diff-d813b91502ac37baa1004104af667747R16